design/open/openapi-multipart-upload.md

```diff
@@ -0,0 +1,217 @@
+    parameters:
+      - in: path
+        name: repository
+        required: true
+        schema:
+          type: string
+      - in: path
+        name: branch
+        required: true
+        schema:
+          type: string
+      - in: query
+        name: path
+        description: relative to the branch
+        required: true
+        schema:
+          type: string
+      - in: query
+        name: parts
+        description: number of parts required to upload
+        schema:
+          type: integer
+          minimum: 1
+          maximum: 1000
+    post:
+      tags:
+        - staging
+      operationId: createMultipartUpload
+      summary: Initiate a multipart upload
+      description: Initiates a multipart upload and returns an upload ID with presigned URLs for each part. Part numbers starts with 1 (limit to 1000 parts). Each part minimum size is 5M.
+      responses:
+        200:
+          description: Multipart upload initiated
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/MultipartUpload"
+        # ... other error responses
+
+  /repositories/{repository}/branches/{branch}/staging/multipart/{uploadId}:
+      parameters:
+      - in: path
+        name: repository
+        required: true
+        schema:
+          type: string
+      - in: path
+        name: branch
+        required: true
+        schema:
+          type: string
+      - in: query
+        name: path
+        description: relative to the branch
+        required: true
+        schema:
+          type: string
+    put:
+      tags:
+        - staging
+      operationId: completeMultipartUpload
+      summary: Complete a multipart upload
+      description: Completes a multipart upload by assembling the uploaded parts.
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/CompleteMultipartUpload"
+      responses:
+        200:
+          description: Multipart upload completed
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ObjectStats"
+        # ... other error responses
+
+    delete:
+      tags:
+        - staging
+      operationId: abortMultipartUpload
+      summary: Abort a multipart upload
+      description: Aborts a multipart upload.
+      responses:
+        204:
+          description: Multipart upload aborted
+
+```
+
+
+**Schemas used by operations**
+
+```yaml=
+components:
+  schemas:
+    MultipartUpload:
+      type: object
+      properties:
+        upload_id:
+          type: string
+        locations:
+          type: array
+          items:
+            $ref: "#/components/schemas/StagingLocation"
+      required:
+        - upload_id
+        
+    UploadPart:
+      type: object
+      properties:
+        part_number:
+          type: integer
+          minimum: 1
+          maximum: 1000
```

Comment:
S3 allows [10_000 parts](https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html).  In any case I do not believe that this limitation should be in the OpenAPI spec: if I later decide to change it on lakeFS, old clients should be able to use this higher maximum value.